src/Swarm/TUI/Model.hs

```diff
@@ -302,6 +312,59 @@ getCurrentItemText history = replItemText <$> Seq.lookup (history ^. replIndex)
 replIndexIsAtInput :: REPLHistory -> Bool
 replIndexIsAtInput repl = repl ^. replIndex == replLength repl
 
+------------------------------------------------------------
+-- Repl Prompt
+------------------------------------------------------------
+
+-- | This data type represent what is prompted to the player
+--   and how the REPL show interpret the user input.
+data REPLPrompt
+  = -- | Interpret the given text as a regular command
+    CmdPrompt Text
+  | -- | Interpret the given text as "search this text in history"
+    SearchPrompt Text REPLHistory
+
+-- | Get the last Entry in REPLHistory matching the given text
+lastEntry :: Text -> REPLHistory -> Maybe Text
+lastEntry t h = case filter (liftA2 (&&) matchesText isREPLEntry) $ toList $ h ^. replSeq of
+  [] -> Nothing
+  rh : _ -> Just $ replItemText rh
+ where
+  matchesText histItem = t `T.isInfixOf` replItemText histItem
+
+-- | The default REPL prompt.
+defaultPrompt :: REPLPrompt
+defaultPrompt = CmdPrompt ""
+
+-- | Lens for accesing the text of the prompt.
+promptTextL :: Lens' REPLPrompt Text
+promptTextL = lens g s
+ where
+  -- Notice that the prompt ADT must have a Text field in every constructor (representing what the user writes).
+  -- This should be force in the ADT itself... right know this here
+  -- The compiler will complain about "Non complete patterns" on this two function.
+  g :: REPLPrompt -> Text
+  g (CmdPrompt t) = t
+  g (SearchPrompt t _) = t
+
+  s :: REPLPrompt -> Text -> REPLPrompt
+  s (CmdPrompt _) t = CmdPrompt t
+  s (SearchPrompt _ h) t = SearchPrompt t h
+
+-- | Turn the repl prompt into a decorator for the form
+replPromptAsWidget :: REPLPrompt -> Widget Name
+replPromptAsWidget (CmdPrompt _) = txt "> "
+replPromptAsWidget (SearchPrompt t rh) =
+  case lastEntry t rh of
+    Nothing -> txt $ "[find " <> t <> " nothing found] "
+    Just lastentry
+      | T.null t -> txt $ "[find " <> t <> "] "
+      | otherwise -> txt $ "[find " <> t <> ": \"" <> lastentry <> "\"] "
+
+-- | Creates the repl form as a decorated form.
+mkReplForm :: REPLPrompt -> Form REPLPrompt AppEvent Name
+mkReplForm r = newForm [(replPromptAsWidget r <+>) @@= editTextField promptTextL REPLInput (Just 1)] r
```

Comment:
Perhaps we should not repeat the search input, it is already displayed in the `[find XX]`. For example, in bash:

- `$` <kbd>CTRL-R</kdb>
- `(reverse-i-search)``':` <kbd>swarm</kbd>
- `(reverse-i-search``swarm': ./dist-newstyle/build/...`

Reply:
Yes, the thing happening is that the form display the text, and also does the decorator. Should be only one of both